saleor/plugins/openid_connect/utils.py

```diff
@@ -0,0 +1,571 @@
+from authlib.oidc.core import CodeIDToken
+from django.contrib.auth.models import Permission
+from django.core.cache import cache
+from django.core.exceptions import ValidationError
+from django.core.validators import URLValidator
+from django.db.models import QuerySet
+from django.middleware.csrf import _compare_masked_tokens  # type: ignore
+from django.middleware.csrf import _get_new_csrf_token
+from django.utils.timezone import make_aware
+from jwt import PyJWTError
+
+from saleor.account.models import User
+from saleor.core.jwt import (
+    JWT_ACCESS_TYPE,
+    JWT_OWNER_FIELD,
+    JWT_REFRESH_TYPE,
+    PERMISSIONS_FIELD,
+    jwt_decode,
+    jwt_encode,
+    jwt_user_payload,
+)
+from saleor.core.permissions import get_permission_names, get_permissions_from_codenames
+from saleor.core.utils.url import validate_storefront_url
+from saleor.plugins.error_codes import PluginErrorCode
+from saleor.plugins.models import PluginConfiguration
+
+from . import PLUGIN_ID
+from .exceptions import AuthenticationError
+
+JWKS_KEY = "oauth_jwks"
+JWKS_CACHE_TIME = 60 * 60  # 1 hour
+USER_INFO_DEFAULT_CACHE_TIME = 60 * 60  # 1 hour
+
+REQUEST_TIMEOUT = 5
+
+
+OAUTH_TOKEN_REFRESH_FIELD = "oauth_refresh_token"
+CSRF_FIELD = "csrf_token"
+
+
+logger = logging.getLogger(__name__)
+
+
+def validate_storefront_redirect_url(storefront_redirect_url: Optional[str]):
+    if not storefront_redirect_url:
+        raise ValidationError(
+            {
+                "redirectUrl": ValidationError(
+                    "Missing redirect url.", code=PluginErrorCode.NOT_FOUND.value
+                )
+            }
+        )
+    try:
+        validate_storefront_url(storefront_redirect_url)
+    except ValidationError as error:
+        raise ValidationError(
+            {"redirectUrl": error}, code=PluginErrorCode.INVALID.value
+        )
+
+
+def fetch_jwks(jwks_url) -> Optional[dict]:
+    response = None
+    try:
+        response = requests.get(jwks_url, timeout=REQUEST_TIMEOUT)
+        response.raise_for_status()
+        jwks = response.json()
+    except requests.exceptions.RequestException:
+        logger.exception("Unable to fetch jwks from %s", jwks_url)
+        raise AuthenticationError("Unable to finalize the authentication process.")
+    except json.JSONDecodeError:
+        content = response.content if response else "Unable to find the response"
+        logger.exception(
+            "Unable to decode the response from auth service with jwks. "
+            "Response: %s",
+            content,
+        )
+        raise AuthenticationError("Unable to finalize the authentication process.")
+    keys = jwks.get("keys", [])
+    if not keys:
+        logger.warning("List of JWKS keys is empty")
+    return keys
+
+
+def get_jwks_keys_from_cache_or_fetch(jwks_url: str) -> dict:
+    jwks_keys = cache.get(JWKS_KEY)
+    if jwks_keys is None:
+        jwks_keys = fetch_jwks(jwks_url)
+        cache.set(JWKS_KEY, jwks_keys, JWKS_CACHE_TIME)
```

Comment:
I don't think it's good place to set keys into cache. `fetch_jwsk(url, cache=True)` might be better. I think getter method should not setting anything.